File: bsp/boot/riscv/arch/elf_reloc.c
/*
 * elf_reloc.c - RISC-V ELF relocation support
 */

#include <sys/param.h>
#include <boot.h>
#include <sys/elf.h>

#define R_RISCV_NONE            0
#define R_RISCV_32              1
#define R_RISCV_64              2
#define R_RISCV_RELATIVE        3
#define R_RISCV_COPY            4
#define R_RISCV_JUMP_SLOT       5
#define R_RISCV_BRANCH          16
#define R_RISCV_JAL             17
#define R_RISCV_CALL            18
#define R_RISCV_CALL_PLT        19
#define R_RISCV_GOT_HI20        20
#define R_RISCV_PCREL_HI20      23
#define R_RISCV_PCREL_LO12_I    24
#define R_RISCV_PCREL_LO12_S    25
#define R_RISCV_HI20            26
#define R_RISCV_LO12_I          27
#define R_RISCV_LO12_S          28
#define R_RISCV_TPREL_HI20      29
#define R_RISCV_TPREL_LO12_I    30
#define R_RISCV_TPREL_LO12_S    31
#define R_RISCV_TPREL_ADD       32
#define R_RISCV_ADD8            33
#define R_RISCV_ADD16           34
#define R_RISCV_ADD32           35
#define R_RISCV_ADD64           36
#define R_RISCV_SUB8            37
#define R_RISCV_SUB16           38
#define R_RISCV_SUB32           39
#define R_RISCV_SUB64           40
#define R_RISCV_GNU_VTINHERIT   41
#define R_RISCV_GNU_VTENTRY     42
#define R_RISCV_ALIGN           43
#define R_RISCV_RVC_BRANCH      44
#define R_RISCV_RVC_JUMP        45
#define R_RISCV_RVC_LUI         46
#define R_RISCV_GPREL_I         47
#define R_RISCV_GPREL_S         48
#define R_RISCV_TPREL_I         49
#define R_RISCV_TPREL_S         50
#define R_RISCV_RELAX           51
#define R_RISCV_SUB6            52
#define R_RISCV_SET6            53
#define R_RISCV_SET8            54
#define R_RISCV_SET16           55
#define R_RISCV_SET32           56
#define R_RISCV_32_PCREL        57

/*
 * A small LUT to store the calculated offset of HI20 relocations,
 * so they can be retrieved by the following LO12 relocations.
 */
#define MAX_HI20 256
static struct {
    Elf32_Addr addr;
    int32_t offset;
} hi20_lut[MAX_HI20];
static int hi20_idx = 0;

static void add_hi20(Elf32_Addr addr, int32_t offset)
{
    hi20_lut[hi20_idx].addr = addr;
    hi20_lut[hi20_idx].offset = offset;
    hi20_idx = (hi20_idx + 1) % MAX_HI20;
}

static int32_t find_hi20(Elf32_Addr addr)
{
    int i;
    for (i = 0; i < MAX_HI20; i++) {
        if (hi20_lut[i].addr == addr) {
            int32_t off = hi20_lut[i].offset;
            hi20_lut[i].addr = 0; /* Clear it */
            return off;
        }
    }
    return 0;
}

int relocate_rela(Elf32_Rela* rela, Elf32_Addr sym_val, char* target_sect)
{
    Elf32_Addr* where;
    Elf32_Addr val;
    int32_t offset;
    uint32_t hi, lo;
    int type = (int)ELF32_R_TYPE(rela->r_info);

    where = (Elf32_Addr*)(target_sect + rela->r_offset);
    val = sym_val + rela->r_addend;

    switch (type) {
    case R_RISCV_NONE:
    case R_RISCV_RELAX:
    case R_RISCV_ALIGN:
        break;

    case R_RISCV_32:
        *where = val;
        break;

    case R_RISCV_HI20:
        /* LUI type */
        offset = (int32_t)val;
        hi = (uint32_t)(offset + 0x800) >> 12;
        *where = (*where & 0x00000fff) | (hi << 12);
        break;

    case R_RISCV_LO12_I:
        /* ADDI type */
        offset = (int32_t)val;
        lo = (uint32_t)offset & 0xfff;
        *where = (*where & 0x000fffff) | (lo << 20);
        break;

    case R_RISCV_LO12_S:
        /* SW type */
        offset = (int32_t)val;
        lo = (uint32_t)offset & 0xfff;
        *where = (*where & 0x01fff07f) | ((lo & 0xfe0) << 20) | ((lo & 0x01f) << 7);
        break;

    case R_RISCV_PCREL_HI20:
        /* AUIPC type */
        offset = (int32_t)val - (int32_t)where;
        hi = (uint32_t)(offset + 0x800) >> 12;
        *where = (*where & 0x00000fff) | (hi << 12);
        add_hi20((Elf32_Addr)where, offset);
        break;

    case R_RISCV_PCREL_LO12_I:
        /* Retrieve offset from LUT */
        offset = find_hi20(sym_val);
        lo = (uint32_t)offset & 0xfff;
        *where = (*where & 0x000fffff) | (lo << 20);
        break;

    case R_RISCV_PCREL_LO12_S:
        /* Retrieve offset from LUT */
        offset = find_hi20(sym_val);
        lo = (uint32_t)offset & 0xfff;
        *where = (*where & 0x01fff07f) | ((lo & 0xfe0) << 20) | ((lo & 0x01f) << 7);
        break;

    case R_RISCV_CALL:
    case R_RISCV_CALL_PLT:
        /* AUIPC + JALR */
        offset = (int32_t)val - (int32_t)where;
        hi = (uint32_t)(offset + 0x800) >> 12;
        lo = (uint32_t)offset & 0xfff;
        DPRINTF(("Reloc CALL at %lx to %lx (off=%lx, hi=%x, lo=%x)\n", (long)where, (long)val, (long)offset, hi, lo));
        /* Patch auipc */
        *where = (*where & 0x00000fff) | (hi << 12);
        /* Patch jalr */
        *(where + 1) = (*(where + 1) & 0x000fffff) | (lo << 20);
        break;

    case R_RISCV_BRANCH:
        offset = (int32_t)val - (int32_t)where;
        *where = (*where & 0x01fff07f) |
                 (((offset >> 12) & 0x01) << 31) |
                 (((offset >> 5) & 0x3f) << 25) |
                 (((offset >> 1) & 0x0f) << 8) |
                 (((offset >> 11) & 0x01) << 7);
        break;

    case R_RISCV_JAL:
        offset = (int32_t)val - (int32_t)where;
        *where = (*where & 0x0000007f) |
                 (((offset >> 20) & 0x01) << 31) |
                 (((offset >> 1) & 0x3ff) << 21) |
                 (((offset >> 11) & 0x01) << 20) |
                 (((offset >> 12) & 0xff) << 12);
        break;

    case R_RISCV_ADD32:
        *where += val;
        break;

    case R_RISCV_SUB32:
        *where -= val;
        break;

    case R_RISCV_32_PCREL:
        *where = val - (Elf32_Addr)where;
        break;

    case R_RISCV_SUB8:
        *(uint8_t *)where = (*(uint8_t *)where) - (uint8_t)(val);
        break;

    case R_RISCV_SUB16:
        *(uint16_t *)where = (*(uint16_t *)where) - (uint16_t)(val);
        break;

    case R_RISCV_SUB6:
        *(uint8_t *)where = (*(uint8_t *)where & 0xc0) | (((*(uint8_t *)where & 0x3f) - (val & 0x3f)) & 0x3f);
        break;

    case R_RISCV_SET6:
        *(uint8_t *)where = (*(uint8_t *)where & 0xc0) | (val & 0x3f);
        break;

    case R_RISCV_SET8:
        *(uint8_t *)where = (uint8_t)val;
        break;

    case R_RISCV_SET16:
        *(uint16_t *)where = (uint16_t)val;
        break;

    case R_RISCV_SET32:
        *(uint32_t *)where = (uint32_t)val;
        break;

    default:
        DPRINTF(("RISCV-BOOT: Unknown reloc type %d at %lx sym_val=%lx\n", type, (long)where, (long)sym_val));
        return -1;
    }
    return 0;
}

int relocate_rel(Elf32_Rel* rel, Elf32_Addr sym_val, char* target_sect)
{
    return -1;
}

File: bsp/boot/riscv/arch/elf_reloc.zig
// SPDX-License-Identifier: BSD-2-Clause
//
// Copyright (c) 2026, Champ Yen <champ.yen@gmail.com>
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
// are met:
// 1. Redistributions of source code must retain the above copyright
//    notice, this list of conditions and the following disclaimer.
// 2. Redistributions in binary form must reproduce the above copyright
//    notice, this list of conditions and the following disclaimer in the
//    documentation and/or other materials provided with the distribution.
//
// THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
// OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
// SUCH DAMAGE.

// bsp/boot/riscv/arch/elf_reloc.zig — RISC-V ELF relocation engine.
// Replaces bsp/boot/riscv/arch/elf_reloc.c.

const ffi = @import("ffi");
const elf = ffi.elf;
const cfg = ffi.cfg;
const boot = ffi.boot;

inline fn DPRINTF(comptime format: []const u8, args: anytype) void {
    if (cfg.DEBUG) {
        ffi.print(format, args);
    }
}

const RelInfo = packed struct(u32) {
    type: u8,
    sym: u24,
};

const RiscvBranchInstruction = packed struct(u32) {
    opcode: u7,
    imm11: u1,
    imm4_1: u4,
    funct3: u3,
    rs1: u5,
    rs2: u5,
    imm10_5: u6,
    imm12: u1,
};

const RiscvJalInstruction = packed struct(u32) {
    opcode: u7,
    rd: u5,
    imm19_12: u8,
    imm11: u1,
    imm10_1: u10,
    imm20: u1,
};

const RiscvBranchOffset = packed struct(u32) {
    _pad: u1 = 0,
    imm4_1: u4,
    imm10_5: u6,
    imm11: u1,
    imm12: u1,
    sign_extension: u19,
};

const RiscvJalOffset = packed struct(u32) {
    _pad: u1 = 0,
    imm10_1: u10,
    imm11: u1,
    imm19_12: u8,
    imm20: u1,
    sign_extension: u11,
};

// A small LUT to store the calculated offset of HI20 relocations,
// so they can be retrieved by the following LO12 relocations.
const MAX_HI20 = 256;
var hi20_lut = [_]struct {
    addr: elf.types.Addr,
    offset: i32,
}{.{ .addr = 0, .offset = 0 }} ** MAX_HI20;
var hi20_idx: usize = 0;

fn add_hi20(addr: elf.types.Addr, offset: i32) void {
    hi20_lut[hi20_idx].addr = addr;
    hi20_lut[hi20_idx].offset = offset;
    hi20_idx = (hi20_idx + 1) % MAX_HI20;
}

fn find_hi20(addr: elf.types.Addr) i32 {
    var i: usize = 0;
    while (i < MAX_HI20) : (i += 1) {
        if (hi20_lut[i].addr == addr) {
            const off = hi20_lut[i].offset;
            hi20_lut[i].addr = 0; // Clear it
            return off;
        }
    }
    return 0;
}

const RiscvAddressOffset = packed struct(u32) {
    lo12: i12,
    hi20: i20,
};

const RiscvUTypeInstruction = packed struct(u32) {
    opcode: u7,
    rd: u5,
    imm20: u20,
};

const RiscvITypeInstruction = packed struct(u32) {
    opcode: u7,
    rd: u5,
    funct3: u3,
    rs1: u5,
    imm12: i12,
};

const RiscvSTypeInstruction = packed struct(u32) {
    opcode: u7,
    imm4_0: u5,
    funct3: u3,
    rs1: u5,
    rs2: u5,
    imm11_5: u7,
};

const RiscvSTypeOffset = packed struct(i12) {
    imm4_0: u5,
    imm11_5: u7,
};

pub fn relocate_rela(
    rela: *elf.types.Rela,
    sym_val: elf.types.Addr,
    target_sect: [*]u8,
) c_int {
    const where: [*]align(1) elf.types.Addr = @ptrCast(target_sect + rela.r_offset);
    const val: elf.types.Addr = sym_val + @as(elf.types.Addr, @bitCast(rela.r_addend));
    const r_type = @as(RelInfo, @bitCast(rela.r_info)).type;

    switch (r_type) {
        elf.riscv.R_RISCV_NONE, elf.riscv.R_RISCV_RELAX, elf.riscv.R_RISCV_ALIGN => {},

        elf.riscv.R_RISCV_32 => {
            where[0] = val;
        },

        elf.riscv.R_RISCV_HI20 => {
            const offset: i32 = @bitCast(val);
            const off = @as(RiscvAddressOffset, @bitCast(offset));
            const hi_val: i20 = if (off.lo12 < 0) off.hi20 + 1 else off.hi20;
            
            const inst: RiscvUTypeInstruction = @bitCast(where[0]);
            var new_inst = inst;
            new_inst.imm20 = @bitCast(hi_val);
            where[0] = @bitCast(new_inst);
        },

        elf.riscv.R_RISCV_LO12_I => {
            const offset: i32 = @bitCast(val);
            const off = @as(RiscvAddressOffset, @bitCast(offset));
            
            const inst: RiscvITypeInstruction = @bitCast(where[0]);
            var new_inst = inst;
            new_inst.imm12 = @bitCast(off.lo12);
            where[0] = @bitCast(new_inst);
        },

        elf.riscv.R_RISCV_LO12_S => {
            const offset: i32 = @bitCast(val);
            const off = @as(RiscvAddressOffset, @bitCast(offset));
            const s_off = @as(RiscvSTypeOffset, @bitCast(off.lo12));
            
            const inst: RiscvSTypeInstruction = @bitCast(where[0]);
            var new_inst = inst;
            new_inst.imm4_0 = s_off.imm4_0;
            new_inst.imm11_5 = s_off.imm11_5;
            where[0] = @bitCast(new_inst);
        },

        elf.riscv.R_RISCV_PCREL_HI20 => {
            const offset: i32 = @bitCast(val - @intFromPtr(where));
            const off = @as(RiscvAddressOffset, @bitCast(offset));
            const hi_val: i20 = if (off.lo12 < 0) off.hi20 + 1 else off.hi20;
            
            const inst: RiscvUTypeInstruction = @bitCast(where[0]);
            var new_inst = inst;
            new_inst.imm20 = @bitCast(hi_val);
            where[0] = @bitCast(new_inst);
            add_hi20(@intFromPtr(where), offset);
        },

        elf.riscv.R_RISCV_PCREL_LO12_I => {
            const offset = find_hi20(sym_val);
            const off = @as(RiscvAddressOffset, @bitCast(offset));
            
            const inst: RiscvITypeInstruction = @bitCast(where[0]);
            var new_inst = inst;
            new_inst.imm12 = @bitCast(off.lo12);
            where[0] = @bitCast(new_inst);
        },

        elf.riscv.R_RISCV_PCREL_LO12_S => {
            const offset = find_hi20(sym_val);
            const off = @as(RiscvAddressOffset, @bitCast(offset));
            const s_off = @as(RiscvSTypeOffset, @bitCast(off.lo12));
            
            const inst: RiscvSTypeInstruction = @bitCast(where[0]);
            var new_inst = inst;
            new_inst.imm4_0 = s_off.imm4_0;
            new_inst.imm11_5 = s_off.imm11_5;
            where[0] = @bitCast(new_inst);
        },

        elf.riscv.R_RISCV_CALL, elf.riscv.R_RISCV_CALL_PLT => {
            const offset: i32 = @bitCast(val - @intFromPtr(where));
            const off = @as(RiscvAddressOffset, @bitCast(offset));
            const hi_val: i20 = if (off.lo12 < 0) off.hi20 + 1 else off.hi20;
            
            // Patch auipc
            const auipc_inst: RiscvUTypeInstruction = @bitCast(where[0]);
            var new_auipc = auipc_inst;
            new_auipc.imm20 = @bitCast(hi_val);
            where[0] = @bitCast(new_auipc);
            
            // Patch jalr
            const jalr_inst: RiscvITypeInstruction = @bitCast(where[1]);
            var new_jalr = jalr_inst;
            new_jalr.imm12 = @bitCast(off.lo12);
            where[1] = @bitCast(new_jalr);
        },

        elf.riscv.R_RISCV_BRANCH => {
            const offset: i32 = @bitCast(val - @intFromPtr(where));
            const off_struct = @as(RiscvBranchOffset, @bitCast(offset));
            var inst = @as(RiscvBranchInstruction, @bitCast(where[0]));
            inst.imm12 = off_struct.imm12;
            inst.imm10_5 = off_struct.imm10_5;
            inst.imm4_1 = off_struct.imm4_1;
            inst.imm11 = off_struct.imm11;
            where[0] = @bitCast(inst);
        },

        elf.riscv.R_RISCV_JAL => {
            const offset: i32 = @bitCast(val - @intFromPtr(where));
            const off_struct = @as(RiscvJalOffset, @bitCast(offset));
            var inst = @as(RiscvJalInstruction, @bitCast(where[0]));
            inst.imm20 = off_struct.imm20;
            inst.imm10_1 = off_struct.imm10_1;
            inst.imm11 = off_struct.imm11;
            inst.imm19_12 = off_struct.imm19_12;
            where[0] = @bitCast(inst);
        },

        elf.riscv.R_RISCV_ADD32 => {
            where[0] += val;
        },

        elf.riscv.R_RISCV_SUB32 => {
            where[0] -= val;
        },

        elf.riscv.R_RISCV_32_PCREL => {
            where[0] = val - @intFromPtr(where);
        },

        elf.riscv.R_RISCV_SUB8 => {
            const ptr: *u8 = @ptrCast(where);
            ptr.* = @bitCast(@as(i8, @bitCast(ptr.*)) - @as(i8, @intCast(val)));
        },

        elf.riscv.R_RISCV_SUB16 => {
            const ptr: *align(1) u16 = @ptrCast(where);
            ptr.* = @bitCast(@as(i16, @bitCast(ptr.*)) - @as(i16, @intCast(val)));
        },

        elf.riscv.R_RISCV_SUB6 => {
            const ptr: *u8 = @ptrCast(where);
            ptr.* = (ptr.* & 0xc0) | @as(u8, @bitCast((@as(i8, @bitCast(ptr.* & 0x3f)) - @as(i8, @intCast(val & 0x3f))) & 0x3f));
        },

        elf.riscv.R_RISCV_SET6 => {
            const ptr: *u8 = @ptrCast(where);
            ptr.* = (ptr.* & 0xc0) | @as(u8, @intCast(val & 0x3f));
        },

        elf.riscv.R_RISCV_SET8 => {
            const ptr: *u8 = @ptrCast(where);
            ptr.* = @intCast(val);
        },

        elf.riscv.R_RISCV_SET16 => {
            const ptr: *align(1) u16 = @ptrCast(where);
            ptr.* = @intCast(val);
        },

        elf.riscv.R_RISCV_SET32 => {
            where[0] = val;
        },

        else => {
            DPRINTF("RISCV-BOOT: Unknown reloc type {d} at {x} sym_val={x}\n", .{ r_type, @intFromPtr(where), sym_val });
            return -1;
        },
    }
    return 0;
}

pub fn relocate_rel(
    _: [*]elf.types.Rel,
    _: elf.types.Addr,
    _: [*]u8,
) c_int {
    return -1;
}

